### DISCUSSION TITLE
Draw on top of an image

### DISCUSSION THREAD

**Discussion by tinloaf** [emilk/egui#2967] [Q&A]
Hi,

First of all: I'm new to egui, in fact I'm new to Rust. So please excuse if this question is stupid and/or asked in a stupid way. ;-)

I want to achieve the following: I need to display an image, which I already have coaxed into a `egui_extras::RetainedImage`. Now I want to draw some stuff onto that image, like semi-transparent rectangles, circles, etc. - stuff for which I would usually use a canvas.

I have already learned that I can use [the `Ui::put()` method](https://docs.rs/egui/latest/egui/struct.Ui.html#method.put) to stack things on top of each other, so I feel like I should call something like `my_image.put(the_correct_max_rect, some_transparent_canvas)`. However, the only way I see to get a canvas is via `Frame::canvas()`, which does not result in a `Widget` that I can hand over to `Ui::put()`.

Can I somehow coax a `Frame::canvas()` into `Ui::put()`? Or is there a better way of achieve drawing on an image?

**Comment by grantshandy** [comment]
Hi @tinloaf, after a bit of research I think I have a way that you can achieve the picture-painting idea. `Frame::canvas`  returns a `Frame` that has a "canvas style", it doesn't return an HTML-like "canvas" object. The first thing I'd look at is the [painting example](https://github.com/emilk/egui/blob/master/crates/egui_demo_lib/src/demo/painting.rs) in the example demo which does something similar to what I imagine you want to do (user drawing shapes).

If you look through the source you'll find that the main part driving the line drawing is [`egui::Painter`](https://docs.rs/egui/0.21.0/egui/struct.Painter.html#method.image) which you can get from your [`Ui::allocate_painter`](https://docs.rs/egui/0.21.0/egui/struct.Ui.html#method.allocate_painter). `Painter` provides you `Painter::circle`, `Painter::rect` and `Painter::image` all inline rather than overlaying a drawing space on top of a retained image directly. The painter example provides some insight into how to accept user input and store/redraw the state. In your case you can draw the image from it's stored ID and then draw the user-created shapes on top afterwards.

Hope that helps! 👍🏽
